compiler/rustc_passes/src/dead.rs

```diff
@@ -244,7 +244,7 @@ impl<'tcx> MarkSymbolVisitor<'tcx> {
             }
 
             // Avoid accessing the HIR for the synthesized associated type generated for RPITITs.
-            if self.tcx.opt_rpitit_info(id).is_some() {
+            if self.tcx.opt_rpitit_info(id.to_def_id()).is_some() {
```

Comment:
I'm not sure what's makes the `opt_rpitit_info` query be fired that much, if it's this call or the other one (because there are just 2), but this call could be the one. I was wondering, what if we make `find_by_def_id` called below return `None` for rpitits and the code on line 256 would just be ...

```rust
self.live_symbols.insert(id);

if let Some(node) = self.tcx.hir().find_by_def_id(id) {
    self.visit_node(node);
}
```

I'm not sure if there's any other situation where `find_by_def_id` returns `None` and we don't want to do `self.live_symbols.insert(id)`, though.

Reply:
With this patch ...

```diff
diff --git a/compiler/rustc_middle/src/hir/map/mod.rs b/compiler/rustc_middle/src/hir/map/mod.rs
index 746cf488589..e76de89e432 100644
--- a/compiler/rustc_middle/src/hir/map/mod.rs
+++ b/compiler/rustc_middle/src/hir/map/mod.rs
@@ -316,7 +316,7 @@ pub fn find(self, id: HirId) -> Option<Node<'hir>> {
     /// Retrieves the `Node` corresponding to `id`, returning `None` if cannot be found.
     #[inline]
     pub fn find_by_def_id(self, id: LocalDefId) -> Option<Node<'hir>> {
-        self.find(self.local_def_id_to_hir_id(id))
+        self.find(self.tcx.opt_local_def_id_to_hir_id(id)?)
     }

     /// Retrieves the `Node` corresponding to `id`, panicking if it cannot be found.
diff --git a/compiler/rustc_passes/src/dead.rs b/compiler/rustc_passes/src/dead.rs
index 73ee51d5f3a..1d713bbd055 100644
--- a/compiler/rustc_passes/src/dead.rs
+++ b/compiler/rustc_passes/src/dead.rs
@@ -243,18 +243,12 @@ fn mark_live_symbols(&mut self) {
                 continue;
             }

-            // Avoid accessing the HIR for the synthesized associated type generated for RPITITs.
-            if self.tcx.opt_rpitit_info(id).is_some() {
-                self.live_symbols.insert(id);
-                continue;
-            }
-
             // in the case of tuple struct constructors we want to check the item, not the generated
             // tuple struct constructor function
             let id = self.struct_constructors.get(&id).copied().unwrap_or(id);

+            self.live_symbols.insert(id);
             if let Some(node) = self.tcx.hir().find_by_def_id(id) {
-                self.live_symbols.insert(id);
                 self.visit_node(node);
             }
         }
```

all the tests pass. May worth running perf on this?.

Reply:
To be clear, if we keep `opt_rpitit_info` around, other PRs may add other calls to the query that may cause regressions. Let's avoid this by fixing the problem at its source. This optimization may also be helpful, feel free to test it if you want, but I'm gonna land this PR I think.

>  if it's this call or the other one (because there are just 2)

There are 5 calls to `opt_rpitit_info` that I can see in the codebase.